src/query/api/v1/handler/prometheus/remote/write_test.go

```diff
@@ -418,6 +422,54 @@ func TestPromWriteDisabledMetricsTypes(t *testing.T) {
 	require.NoError(t, capturedIter.Error())
 }
 
+func TestPromWriteLiteralIsTooLongError(t *testing.T) {
+	ctrl := xtest.NewController(t)
+	defer ctrl.Finish()
+
+	observerCore, observedLogs := observer.New(zapcore.InfoLevel)
+	opts := makeOptions(ingest.NewMockDownsamplerAndWriter(ctrl))
+	opts = opts.SetInstrumentOpts(opts.InstrumentOpts().SetLogger(zap.New(observerCore)))
+	handler, err := NewPromWriteHandler(opts)
+	require.NoError(t, err)
+
+	veryLongLiteral := ""
+	for ; len(veryLongLiteral) <= int(opts.TagOptions().MaxTagLiteralLength()); veryLongLiteral += "very_long_literal_" {
+	}
+	promReq := &prompb.WriteRequest{
+		Timeseries: []prompb.TimeSeries{
+			{
+				Labels: []prompb.Label{
+					{Name: []byte("name1"), Value: []byte("value1")},
+					{Name: []byte("name2"), Value: []byte(veryLongLiteral)},
+				},
+			},
+		},
+	}
+
+	// execute write request many times to check that "literal is too long" sampling doesn't error
+	// or deadlock
+	for i := 0; i < maxLiteralIsTooLongLogCount*2; i++ {
+		promReqBody := test.GeneratePromWriteRequestBody(t, promReq)
+		req := httptest.NewRequest(PromWriteHTTPMethod, PromWriteURL, promReqBody)
+		handler.ServeHTTP(httptest.NewRecorder(), req)
+	}
+	filteredLogs := observedLogs.Filter(func(e observer.LoggedEntry) bool {
+		return e.Message == "label exceeds literal length limits"
+	})
+	require.True(t, filteredLogs.Len() == maxLiteralIsTooLongLogCount,
+		"unexpected number of log messages, expected=%v, actual=%v", maxLiteralIsTooLongLogCount, filteredLogs.Len())
+	for _, msg := range filteredLogs.All() {
+		for _, f := range msg.Context {
+			switch f.Key {
+			case "name":
+				assert.Equal(t, "name2", f.String)
+			case "value":
+				assert.Equal(t, veryLongLiteral[:literalLengthThreshold]+"...", f.String)
+			}
+		}
+	}
```

Comment:
Where do we check response status code or whatever matters for the client?
Usually we don't test logging, I believe this can be omitted.